### PR TITLE
[Refactor][MOSS-TTS] Consolidate the MOSS-Audio-Tokenizer attention backend

### PR DESCRIPTION
**Pull request by CloudRipple**
## Motivation

MOSS-TTS Delay and MOSS-TTS Local execute the same non-streaming MOSS-Audio-Tokenizer local-causal decoder attention, but the repository did not have a single, explicit backend policy for it. Backend selection, packed-sequence preparation, RoPE handling, and model-layout adaptation were coupled in the same implementation. The packed path was also tied to the Hugging Face `flash_attention_2` configuration name, making its behavior difficult to reason about across CUDA versions, dtypes, and GPU architectures.

We compared the available implementations and execution strategies for this workload: FlashInfer FA2, SGLang packed FlashAttention, full-sequence versus 128-token tiled packed execution, and query-chunked SDPA versus HF duration-based streaming. The comparison showed that one backend should not be forced across all environments: packed FlashAttention is the fast path when supported, 128-token tiling is required for the `SM < 90` path, and query-chunked SDPA is the portable fallback. FlashInfer did not provide a demonstrated end-to-end benefit, so it is not added to the runtime policy.

This PR consolidates the shared MOSS-Audio-Tokenizer attention runtime and makes the selection policy explicit and testable. It preserves the decoder math and source weights, supports both v1 and v2 module layouts, and removes redundant Q/K/V contiguous copies.

## Modifications

- Add `sglang_omni/models/moss_tts/attention.py` with the shared MOSS-Audio-Tokenizer attention implementation:
  - explicit `auto`, `packed_flash_attention`, and `sdpa` backend policies;
  - SGLang's `_is_fa3_supported()` predicate for packed-kernel eligibility;
  - BF16-only packed FlashAttention eligibility on CUDA (FP16 is intentionally not added);
  - query-chunked SDPA fallback for CPU, FP32, explicit SDPA, or unavailable packed kernels;
  - per-device caching for the SM-based direct-versus-tiled routing decision;
  - packed/unpacked sequence helpers, cached packed RoPE, and the equivalent local-causal FlashAttention window `(context - 1, 0)`;
  - direct Q/K/V views passed to the packed kernel, without redundant `.contiguous()` copies.
- Simplify `audio_tokenizer.py` into the shared transformer and decoder wrappers plus codec loading:
  - use explicit `from_module()` constructors that reuse the source modules;
  - support both MOSS-Audio-Tokenizer v1 weight fields and v2 `ffn` fields;
  - resolve one consistent attention backend across all transformer layers;
  - use direct packed FA3 on `SM >= 90` and restore a shared 128-token local-window plan with overlapping K/V gathers on `SM < 90` for cross-architecture compatibility.
- Expand the MOSS-Audio-Tokenizer unit coverage for backend policy and fallback, SGLang FA3 capability checks, local-window equivalence, query-chunked SDPA, zero-length and packed/unpacked inputs, v1/v2 layouts, and CUDA BF16 parity.
- Update `tests/README.md` so the shared attention and decoder tests are listed under `unit_test/moss_tts`.

The change is limited to the non-streaming packed decoder path. Delay streaming uses the original codec `model.decode()` path, and MOSS-TTS Local streaming uses the persistent `codec.streaming()` / `_decode_frame()` session. No streaming attention implementation is replaced by this PR.

## Related Issues

#1233 

## Accuracy Test

The following end-to-end A/B on one NVIDIA H200. Each consumer has five base rounds and five PR rounds, with alternating execution order, seed 42, and concurrency 16.

| Item | MOSS-TTS Delay | MOSS-TTS Local |
| --- | ---: | ---: |
| Checkpoint | `OpenMOSS-Team/MOSS-TTS-v1.5` | `OpenMOSS-Team/MOSS-TTS-Local-Transformer-v1.5` |
| Samples per round | 1,088 | 1,088 |
| Base/PR rounds | 5 / 5 | 5 / 5 |
| Concurrency | 16 | 16 |
| Requests (base + PR) | 10,880 | 10,880 |
| Failed requests | 0 | 0 |

Quality values are five-round means. WER is shown as a percentage; its delta is the paired base-to-PR change in percentage points.

| Model | Metric | Base | PR | Delta |
| --- | --- | ---: | ---: | ---: |
| Delay | Corpus WER | 2.0715% | 2.1636% | +0.0921 pp |
| Delay | WavLM speaker similarity | 69.0227 | 68.9378 | -0.0849 |
| Delay | UTMOS | 3.9442 | 3.9443 | +0.0001 |
| Local | Corpus WER | 2.1619% | 2.0916% | -0.0703 pp |
| Local | WavLM speaker similarity | 65.3359 | 65.1872 | -0.1487 |
| Local | UTMOS | 3.9667 | 3.9659 | -0.0008 |

All 21,760 generation requests completed successfully. Every generated WAV was finite, with the expected 24 kHz Delay or 48 kHz Local sample rate. The WER, speaker-similarity, UTMOS, and audio-integrity gates passed for both consumers.

One Delay PR round showed a throughput outlier (`7.826 req/s`). The paired base and PR round was rerun with the same seed, dataset, hardware, and concurrency; the replacement values were `8.848` and `8.872 req/s`. The results below use the replacement run.

A focused H200 long-sequence check repeated codec frames to cross the local-window boundary and produced a finite 64-second, 48 kHz stereo waveform. Manual listening found no audible artifact. A follow-up A800 run exposed severe audio corruption without 128-token tiling, which led to the current `SM < 90` tiled path. The targeted A800 correctness and long-audio gates remain outstanding.

## Benchmark & Profiling

The following results are five-round means from the H200 concurrency-16 base-versus-PR campaign. Speed deltas are paired base-to-PR means; negative latency and RTF deltas are improvements.

| Model | Metric | Base | PR | Delta |
| --- | --- | ---: | ---: | ---: |
| Delay | Throughput (req/s) | 8.6042 | 8.8040 | +2.400% |
| Delay | Mean latency (s) | 1.8546 | 1.8106 | -2.297% |
| Delay | P95 latency (s) | 2.3470 | 2.2734 | -2.933% |
| Delay | Mean RTF | 0.42786 | 0.41750 | -2.349% |
| Delay | Audio throughput (audio-s/s) | 38.5438 | 39.4376 | +2.397% |
| Local | Throughput (req/s) | 11.4686 | 11.8250 | +3.109% |
| Local | Mean latency (s) | 1.3888 | 1.3462 | -3.066% |
| Local | P95 latency (s) | 1.9022 | 1.8660 | -1.900% |
| Local | Mean RTF | 0.32738 | 0.31722 | -3.103% |
| Local | Audio throughput (audio-s/s) | 50.4762 | 52.0458 | +3.110% |

## Checklist

- [x] Format your code according with pre-commit.
- [x] Add unit tests.
- [x] Update documentation / docstrings / example tutorials as needed.
- [x] Provide throughput / latency benchmark results and accuracy evaluation results as needed.
- [ ] For reviewers: If you haven't made any contributions to this PR and are only assisting with merging the main branch, please remove yourself as a co-author when merging the PR.

## CI

CI runs on self-hosted GPU runners and requires a maintainer to add the
`run-ci` label. Once labeled, every subsequent push re-triggers CI as
long as the label remains. Use `/tag-and-rerun-ci higgs` or
`/tag-and-rerun-ci moss` to select a TTS CI model, and
`/tag-and-rerun-ci fun-asr` or `/tag-and-rerun-ci qwen3-asr` to select an ASR
CI model. One selector from each family can be combined, for example
`/tag-and-rerun-ci moss fun-asr`. Draft PRs are skipped even if labeled.

### CODE REVIEW

File: sglang_omni/models/moss_tts/attention.py [hunk 1]
```diff
@@ -0,0 +1,882 @@
...
+from torch import nn
+
+from sglang_omni.models.moss_tts.vocoder_kernels import (
+    apply_exact_interleaved_rope_inplace,
+)
+
+# note (Zhang Yiyang): Bound SDPA fallback memory with query chunks.
+_SDPA_QUERY_CHUNK_SIZE = 512
+# note (Zhang Yiyang): Keep 128-token local-causal tiles for A800 audio stability.
+_LOCAL_CAUSAL_FLASH_QUERY_CHUNK_SIZE = 128
```
Comment (luojiaxuan) [inline sglang_omni/models/moss_tts/attention.py:27]: sglang's `_is_fa3_supported` returns True for sm80/86/89 too (A100/A800/L40/4090), so all these cards go through this tiled path in production. But the H200 A/B and CI (Hopper) only cover the sm90 direct path — this tiled path is only covered by mock-kernel unit tests so far. A800 gates are still outstanding per the PR description. Better to finish them before merge, or at least open an issue to track. The sm<90 path is exactly what this PR is fixing, so it shouldn't be the only unverified one.

File: sglang_omni/models/moss_tts/attention.py [hunk 2]
```diff
@@ -0,0 +1,882 @@
...
+import math
+from collections.abc import Callable, Sequence
+from dataclasses import dataclass
+from functools import cache
+from itertools import accumulate
+
+import torch
+import torch.nn.functional as F
+from sglang.jit_kernel.flash_attention import flash_attn_varlen_func
+from sglang.jit_kernel.flash_attention_v3 import _is_fa3_supported
```
Comment (luojiaxuan) [inline sglang_omni/models/moss_tts/attention.py:17]: `_is_fa3_supported` is private, and upstream sglang already moved this module to `sglang.kernels.ops.attention/`. Fine with the current pin, but will break on next sglang bump. Maybe leave a comment here.

**Review by luojiaxuan — APPROVED** [review]
LG

File: sglang_omni/models/moss_tts/attention.py (hunk 1 above)
Comment (CloudRipple) [reply sglang_omni/models/moss_tts/attention.py:27]: Thanks for pointing this out. The `SM < 90` tiled path is not a new, unvalidated implementation introduced by this PR. It was the existing default MOSS-Audio-Tokenizer path in the parent revision, using 128-token query tiles with overlapping K/V gathers.

During optimization, we tried removing the tiling to improve efficiency and use direct full-sequence packed FlashAttention. Although the direct path worked on H200, the A800 run exposed severe audio corruption. We therefore restored the existing 128-token tiled behavior and made the routing architecture-aware: direct packed FlashAttention for `SM >= 90`, and the established tiled path for `SM < 90`.

This PR only moves the shared implementation into `attention.py` and adds the new direct path for `SM >= 90`; it does not replace the existing compatibility path for A800 and other pre-SM90 devices. The H200 A/B therefore measures the new direct path, while the `SM < 90` path remains the established baseline.

File: sglang_omni/models/moss_tts/attention.py (hunk 2 above)
Comment (CloudRipple) [reply sglang_omni/models/moss_tts/attention.py:17]: Fixed.
